src/lib/core/GroupedCallbackList.h

```diff
@@ -0,0 +1,227 @@
+    //
+    // This allows the start of a group to be reached from any group member via
+    // ->prev->next. Nodes in a group can be enumerated by via the "next" chain,
+    // inspecting the "prev" pointers to detect the end of the group. The price
+    // for encoding grouping in this way is that upon removal of a group we have
+    // to update not just the "prev" pointer of the following node, but of all
+    // nodes in the following group.
+    //
+    // When retrieving a (sparse) group from the list, we also need to be able
+    // to tell which callbacks are present: In a grouped list with types (A, B)
+    // both (a, nullptr) and (nullptr, b) are by necessity represented by only
+    // a single node in the list. To be able to recover this information, we use
+    // distinct trampolines that tag the "cancel" function pointer stored in each
+    // node with the index of the callback type within the argument type tuple.
+
+    template <std::size_t... Index>
+    void Enqueue(std::index_sequence<Index...>, Callback<T> *... callback)
+    {
+        Cancelable * group = nullptr;
+        (
+            [&] {
+                VerifyOrReturn(callback != nullptr);
+                detail::EnqueueWithGroup(callback->Cancel(), group, this, &detail::TaggedDequeueGroup<Index>);
+            }(),
+            ...);
+    }
+
+    template <std::size_t... Index>
+    bool Peek(std::index_sequence<Index...>, Callback<T> *&... callback) const
+    {
+        Cancelable * cancelable = mNext;
+        VerifyOrReturnValue(cancelable != this, false);
+        Cancelable * groupPrev = cancelable->mPrev;
+        (
+            [&] {
+                if (cancelable->mPrev == groupPrev && cancelable->mCancel == &detail::TaggedDequeueGroup<Index>)
+                {
+                    callback   = Callback<decltype(callback->mCall)>::FromCancelable(cancelable);
+                    cancelable = cancelable->mNext;
+                }
+                else
+                {
+                    callback = nullptr;
+                }
+            }(),
+            ...);
+        return true;
+    }
+};
+
+namespace detail {
+
+inline void EnqueueWithGroup(Cancelable * cancelable, Cancelable *& group, Cancelable * anchor, void (*cancelFn)(Cancelable *))
+{
+    cancelable->mCancel = cancelFn;
+    cancelable->mNext   = anchor;
+    if (!group)
+    {
+        group             = cancelable;
+        cancelable->mPrev = anchor->mPrev;
+    }
+    else
+    {
+        cancelable->mPrev = group->mPrev;
+    }
+    anchor->mPrev->mNext = cancelable;
+    anchor->mPrev        = cancelable;
+}
+
+inline void LinkGroup(Cancelable * prev, Cancelable * cancelable)
+{
+    prev->mNext = cancelable;
+
+    Cancelable * groupPrev = cancelable->mPrev;
+    do
+    {
+        cancelable->mPrev = prev;
+        cancelable        = cancelable->mNext;
+    } while (cancelable->mPrev == groupPrev);
+}
+
+inline Cancelable * ClearCancelable(Cancelable * cancelable)
```

Comment:
Should this be called DequeueCancelable?  I guess it's not adjusting any pointers to `cancelable` from its next/prev... so maybe not.

Reply:
Yeah it's specifically not messing with the pointed-to objects